### PR TITLE
feat(api-client): add pre/post request scripts  🤖

### PR DESCRIPTION
**Pull request by marclave**
🚨 new feature alert 🚨 

Scripting in the Scalar API Client : ) it's pure javascript, with access to a `scalar` object, with some namespaces for functionality depending on if you are in the pre or post script sections

**Present**
Pre Request:
- can access the `request` and add headers : ) (do we give access to the AxiosConfig and let users modify or slight helpers like i did here?)

Post Request:
- tests! you can now run tests and expect values : )
- you also get access to the `response` object to check whats going on : )

image

**Future**
- we should add the scalar object to the Codemirror config so we can add code completion + hints
- we should fix javascript script highlighting
- add a `x-scalar-scripts` OpenAPI extension so we can add these on the fly ✨ 
- Add way more test functionality (do we just use `vitest` browser only?
- This is just at the request level, but we can eventually do: Folder, Collection and workspace level : )
- We should also be able to send requests from the pre request script (foundationally request chaining 😏 )

i disabled this for the Modal mode, so we can get this merged and play around with it as we build : ) but it's pretty fun!!

### CODE REVIEW

**Comment by antlio** [comment]
this seems to be missing some styling, need someone to jump? @marclave

File: packages/api-client/src/libs/scalarScripts.ts [hunk 1]
```diff
@@ -0,0 +1,74 @@
+import type { ResponseTest } from '@scalar/oas-utils/entities/workspace/spec'
+import type { AxiosRequestConfig, AxiosResponse } from 'axios'
+
+function getPreRequestScalarObject(config: AxiosRequestConfig) {
```
Comment (geoffgscott) [inline packages/api-client/src/libs/scalarScripts.ts:4]: Before we launch this let's move to vanilla fetch.

File: packages/api-client/src/libs/scalarScripts.ts [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+  const executeScript = new Function('scalar', script)
+
+  const output = executeScript(scalar)
+
+  return { output, config: scalar._config }
+}
+
+function expect(actual: any) {
+  return {
+    toBe: (expected: any) => {
```
Comment (geoffgscott) [inline packages/api-client/src/libs/scalarScripts.ts:36]: Would be nice to have user generated messages here. Should take an optional second argument that is  `(actual: any, expected: any): string`

Comment (geoffgscott) [reply packages/api-client/src/libs/scalarScripts.ts:36]: If its provided then we should run that function to generate the message and fallback to our message

File: packages/api-client/src/libs/scalarScripts.ts [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        throw new Error(`Expected ${actual} to be ${expected}`)
+      }
+    },
+  }
+}
+
+function getPostRequestScriptObject(response?: AxiosResponse) {
+  const testResults: ResponseTest[] = []
+  return {
+    sub: (a: number, b: number) => a - b,
```
Comment (geoffgscott) [inline packages/api-client/src/libs/scalarScripts.ts:47]: All of these return elements need proper doc strings

File: packages/api-client/src/libs/scalarScripts.ts [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+      try {
+        func()
+        testResults.push({ title, passed: true })
+      } catch (e) {
+        testResults.push({ title, passed: false, error: e as string })
+      }
+    },
+    expect,
+    response,
+    _testResults: testResults,
```
Comment (geoffgscott) [inline packages/api-client/src/libs/scalarScripts.ts:58]: underscore is not a pattern we generally use in our codebase.

File: packages/api-client/src/libs/sendRequest.ts [hunk 5]
```diff
@@ -42,6 +48,7 @@ export const sendRequest = async (
   sentTime?: number
   request?: RequestExample
   response?: ResponseInstance
+  testResults?: ResponseTest[]
 }> => {
   let url = rawUrl
```
Comment (geoffgscott) [inline packages/api-client/src/libs/sendRequest.ts:53]: No `let`.

File: packages/api-client/src/libs/sendRequest.ts [hunk 6]
```diff
@@ -176,15 +183,29 @@ export const sendRequest = async (
   // Append new query string to the URL
   url = `${urlWithoutQueryString}${queryString ? '?' + queryString : ''}`
 
-  const config: AxiosRequestConfig = {
+  let config: AxiosRequestConfig = {
```
Comment (geoffgscott) [inline packages/api-client/src/libs/sendRequest.ts:186]: No let. Use object assign, objectMerge or override properties as needed.

File: packages/api-client/src/libs/sendRequest.ts [hunk 7]
```diff
@@ -203,23 +224,51 @@ export const sendRequest = async (
...
+    // We can handle the post request scripts now
+    if (example.postSendScript) {
+      try {
+        const postRequest = executePostRequestScript(
+          example.postSendScript,
+          response,
+        )
+        testResults = postRequest.testResults
+      } catch (error) {
+        console.warn('Post-send script error:', error)
```
Comment (geoffgscott) [inline packages/api-client/src/libs/sendRequest.ts:236]: Some display of these errors to the user would be sweet. Toast?

File: packages/api-client/src/libs/sendRequest.ts [hunk 8]
```diff
@@ -203,23 +224,51 @@ export const sendRequest = async (
...
       sentTime: Date.now(),
       request: example,
       response: {
         ...response,
         duration: Date.now() - startTime,
       },
+      testResults,
     }
   } catch (error) {
     const axiosError = error as AxiosError
```
Comment (geoffgscott) [inline packages/api-client/src/libs/sendRequest.ts:250]: This is a bad typecasts. Its not guaranteed to be an axios error.

**Review by geoffgscott — CHANGES_REQUESTED** [review]
A few changes needed. Awesome start.

**Comment by marclave** [comment]
moving this to draft for now :)

**Comment by marclave** [comment]
well this ticket got stale quick : ) let's re-add it back post workspace entities PR

**Review by bretthowell714-source — APPROVED** [review]

**Comment by dastanko** [comment]
Any specific reason why this feature is not deployed?

File: packages/api-client/src/libs/scalarScripts.ts (hunk 1 above)
Comment (fiverrwixs-cmyk) [reply packages/api-client/src/libs/scalarScripts.ts:4]: hi

**Pull request closed without merging**